src/sys/pthread.rs

```diff
@@ -34,10 +39,351 @@ pub fn pthread_kill<T>(thread: Pthread, signal: T) -> Result<()>
+    /// # Safety
+    /// 
+    /// Attempting to initialize an already initialized mutex results in undefined behavior.
+    pub unsafe fn init(mutex: *mut Mutex, attr: Option<MutexAttr>) -> Result<()> {
+        let attr = match attr {
+            Some(mut x) => &mut x.0,
+            None => std::ptr::null_mut()
+        };
+        Errno::result(libc::pthread_mutex_init((*mutex).0.get(),attr))?;
+        Ok(())
+    }
+    /// Wraps [`libc::pthread_mutex_init`].
+    pub fn new(attr: Option<MutexAttr>) -> Result<Self> {
+        let attr = match attr {
+            Some(mut x) => &mut x.0,
+            None => std::ptr::null_mut()
+        };
+        let init = unsafe {
+            let mut uninit = std::mem::MaybeUninit::<libc::pthread_mutex_t>::uninit();
+            Errno::result(libc::pthread_mutex_init(uninit.as_mut_ptr(),attr))?;
+            uninit.assume_init()
+        };
+        Ok(Self(UnsafeCell::new(init)))
+    }
+    /// Wraps [`libc::pthread_mutex_lock`].
+    /// 
+    /// <https://man7.org/linux/man-pages/man3/pthread_mutex_lock.3p.html>
+    pub fn lock(&self) -> Result<MutexGuard<'_>> {
+        unsafe {
+            Errno::result(libc::pthread_mutex_lock(self.0.get())).map(|_| MutexGuard(self))
+        }
+    }
+    /// Wraps [`libc::pthread_mutex_trylock`].
+    /// 
+    /// <https://man7.org/linux/man-pages/man3/pthread_mutex_lock.3p.html>
+    pub fn try_lock(&self) -> Result<Option<MutexGuard<'_>>> {
+        unsafe {
+            match Errno::result(libc::pthread_mutex_trylock(self.0.get())) {
+                Ok(_) => Ok(Some(MutexGuard(self))),
+                Err(Errno::EBUSY) => Ok(None),
+                Err(err) => Err(err)
+            }
+            
+        }
+    }
+    /// Wraps [`libc::pthread_mutex_unlock`].
+    /// 
+    /// <https://man7.org/linux/man-pages/man3/pthread_mutex_lock.3p.html>
+    /// 
+    /// Prefer unlocking by dropping the [`MutexGuard`] returned by [`Mutex::lock`] or [`Mutex::try_lock`].
+    /// 
+    /// # Safety
+    /// 
+    /// Results in UB if not called from the thread that locked the mutex.
+    pub unsafe fn unlock(&self) -> Result<()> {
+        unsafe {
+            Errno::result(libc::pthread_mutex_unlock(self.0.get())).map(drop)
+        }
+    }
+}
+
+#[cfg(target_os = "linux")]
+unsafe impl Sync for Mutex {}
+
+#[cfg(target_os = "linux")]
+impl std::ops::Drop for Mutex {
+    /// Wraps [`libc::pthread_mutex_destroy`].
+    fn drop(&mut self) {
+        unsafe {
+            Errno::result(libc::pthread_mutex_destroy(self.0.get())).unwrap();
+        }
+    }
+}
+
+/// Mutex guard to prevent unlocking a mutex from a different thread than the thread that locked it.
+#[cfg(target_os = "linux")]
+#[derive(Debug)]
+pub struct MutexGuard<'a>(&'a Mutex);
+
+#[cfg(target_os = "linux")]
+impl MutexGuard<'_> {
+    /// Calls [`Mutex::unlock`].
+    pub fn try_unlock(self) -> Result<()> {
+        // Prevent calling `Self::Drop` which would attempt to unlock twice.
```

Comment:
What happens if try_unlock fails?  Won't the MutexGuard get dropped, with no way to unlock the mutex later?

Reply:
The mutex could then be unlocked with `Mutex::unlock`.

I'm not sure there is any other to handle this that supports getting this error?

Maybe `pub fn try_unlock(self) -> std::result::Result<(), (Self, nix::Error)>`?